includes/class-find-a-space-ajax-handler.php

```diff
@@ -0,0 +1,265 @@
+
+		// Query single building
+		add_action( 'wp_ajax_find_a_space_single_building', array( $this, 'single_building_callback' ) );
+		add_action( 'wp_ajax_nopriv_find_a_space_single_building', array( $this, 'single_building_callback' ) );
+
+		// List rooms (with filters)
+		add_action( 'wp_ajax_find_a_space_rooms', array( $this, 'rooms_callback' ) );
+		add_action( 'wp_ajax_nopriv_find_a_space_rooms', array( $this, 'rooms_callback' ) );
+
+		// Single room
+		add_action( 'wp_ajax_find_a_space_classroom', array( $this, 'single_classroom_callback' ) );
+		add_action( 'wp_ajax_nopriv_find_a_space_classroom', array( $this, 'single_classroom_callback' ) );
+
+		// Single building
+		add_action( 'wp_ajax_find_a_space_building', array( $this, 'single_building_callback' ) );
+		add_action( 'wp_ajax_nopriv_find_a_space_building', array( $this, 'single_building_callback' ) );
+	}
+
+	private function verify_nonce() {
+		return ( isset( $_REQUEST['_nonce'] )
+			&& wp_verify_nonce( $_REQUEST['_nonce'], self::NONCE_KEY ) );
+	}
+
+	public function meta_callback() {
+		if ( ! $this->verify_nonce() ) {
+			return wp_send_json_error( 'Invalid nonce' );
+		}
+
+		$data = $_REQUEST['data'];
+
+		$campus = sanitize_text_field( $data['campus'] ?? null );
+		$formal = rest_sanitize_boolean( $data['formal'] ?? null );
+
+		$params = array(
+			'campus'       => $campus,
+			'formal'       => $formal,
+			'should_cache' => true,
+		);
+
+		$shared_amenities   = $this->airtable_api->get( 'get_shared_amenities', $params );
+		$resources          = $this->airtable_api->get( 'get_resources', $params );
+		$informal_amenities = $this->airtable_api->get( 'get_informal_amenities', $params );
+		$accessibility      = $this->airtable_api->get( 'get_accessibility', $params );
+		$classroom_layouts  = $this->airtable_api->get( 'get_classroom_layouts', $params );
+		$furniture          = $this->airtable_api->get( 'get_furniture', $params );
+
+		$payload = array(
+			'shared_amenities'   => $shared_amenities,
+			'resources'          => $resources,
+			'informal_amenities' => $informal_amenities,
+			'accessibility'      => $accessibility,
+			'classroom_layouts'  => $classroom_layouts,
+			'furniture'          => $furniture,
+		);
+
+		return wp_send_json( array( 'data' => $payload ) );
+	}
+
+	public function buildings_callback() {
+		if ( ! $this->verify_nonce() ) {
+			return wp_send_json_error( 'Invalid nonce' );
+		}
+
+		$data = $_REQUEST['data'];
+
+		$campus = sanitize_text_field( $data['campus'] ?? null );
+		$formal = rest_sanitize_boolean( $data['formal'] ?? null );
+
+		$params = array(
+			'campus'       => $campus,
+			'formal'       => $formal,
+			'should_cache' => true,
+		);
+
+		$data = $this->airtable_api->get( 'get_buildings', $params );
+
+		return wp_send_json( array( 'data' => $data ) );
+	}
+
+	private function get_html_in_selector( string $html, string $selector ) {
```

Comment:
This method gives me the heebie jeebies. Using DomDocument and DomXPath is normally a last resort. I'm not 100% sure what's going on here, could we have some more inline comments and a docblock to describe what the method is doing?

Reply:
Hi Rich,

This method is definitely not ideal. The expectation of the client is to have dynamic pages within the react app find spaces tool. As we are under a bit of a time crunch I am leveraging some functionality from the other Spaces Pages plugin. The Spaces Pages is creating pages for classrooms `/classrooms/{slug}` and buildings `/buildings/{slug}` which are obtained in the backend of this Find Spaces plugin, the html is loaded into DOMDocument, and only the relevant content is returned to React. React is outputting this html after running it through a purifier to prevent XSS attacks (https://github.com/cure53/DOMPurify).

I'll definitely add some more context in the form of documentation.

Reply:
I've added a documentation block here to explain what we're doing and why. I woudl expect that Brendan can add better and more context here once he returns.